File: src/cowtree/publications.py
"""Commit only a checked candidate and recover its exact acknowledgement."""

from dataclasses import dataclass

from pydantic import TypeAdapter

from cowtree.checks import Checks
from cowtree.durable import write_record
from cowtree.errors import CowtreeError, CowtreeErrorCode
from cowtree.exec import CommandRunner
from cowtree.git import GitRepository
from cowtree.leaves import Leaves
from cowtree.metadata import Metadata
from cowtree.metadata_types import Candidate, Grant, Operation, Receipt, Request
from cowtree.projection import GitProjection
from cowtree.views import Views
from cowtree.workspace import Workspace
from cowtree.workspace_types import Leaf, PublicationRecord


@dataclass(frozen=True)
class Publications:
    workspace: Workspace

    def prepare(self, identity: int) -> Candidate:
        with self.workspace.session() as metadata:
            leaf = Leaves(self.workspace).read(identity=identity)
            pending = leaf.pending
            if pending is None or not pending.submitted or pending.aborting:
                raise CowtreeError(
                    CowtreeErrorCode.INVALID_ARGUMENTS, "capture a source request first"
                )
            candidate = metadata.call(
                Operation.PREPARE, {"request": pending.request.model_dump(mode="json")}
            ).decode("candidate", TypeAdapter(Candidate))
            current = Workspace.open(root=self.workspace.root)
            parent = current.nodes.read(identity=current.config.warm_tip)
            published = Views.snapshot(metadata=metadata)
            if published.tip.version != candidate.parent or published.entries != parent.source:
                raise CowtreeError(
                    CowtreeErrorCode.COMMAND_FAILED, "candidate parent needs projection recovery"
                )
            updated = pending.model_copy(
                update={
                    "candidate": candidate,
                    "batch": None,
                    "parent_node": parent.id,
                    "validation": None,
                }
            )
            Leaves(self.workspace).save(leaf=leaf.model_copy(update={"pending": updated}))
        return candidate

    def commit(self, candidate: Candidate) -> Receipt:
        with self.workspace.session() as metadata:
            leaf = Leaves(self.workspace).read(identity=candidate.request.leaf)
            if leaf.pending is None:
                if leaf.last_receipt is None or leaf.last_receipt.request != candidate.request:
                    raise CowtreeError(
                        CowtreeErrorCode.INVALID_ARGUMENTS, "candidate is not pending"
                    )
                receipt = metadata.call(
                    Operation.COMMIT, {"candidate": candidate.model_dump(mode="json")}
                ).decode("receipt", TypeAdapter(Receipt))
                return receipt
            pending = leaf.pending
            if (
                pending.aborting
                or pending.candidate != candidate
                or pending.validation is None
                or pending.validation.candidate != candidate
            ):
                raise CowtreeError(
                    CowtreeErrorCode.INVALID_ARGUMENTS, "exact candidate has not passed validation"
                )
            node = self.workspace.nodes.read(identity=pending.validation.node)
            self.workspace.nodes.verify(node=node)
            if node.source != Checks(self.workspace).manifest(candidate=candidate):
                raise CowtreeError(
                    CowtreeErrorCode.COMMAND_FAILED, "validated source differs from candidate"
                )
            receipt = metadata.call(
                Operation.COMMIT, {"candidate": candidate.model_dump(mode="json")}
            ).decode("receipt", TypeAdapter(Receipt))
            self.finish(metadata=metadata, leaf=leaf, receipt=receipt)
        return receipt

    def finish(self, metadata: Metadata, leaf: Leaf, receipt: Receipt) -> None:
        pending = leaf.pending
        assert pending is not None
        if pending.validation is None or pending.candidate is None or pending.parent_node is None:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "committed request lacks its validation record"
            )
        if receipt.request != pending.request or receipt.root != pending.candidate.root:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "receipt differs from pending publication"
            )
        published = Views.snapshot(metadata=metadata)
        if published.tip.version != receipt.version:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "publication history advanced before recovery"
            )
        node = self.workspace.nodes.read(identity=pending.validation.node)
        if node.source != published.entries:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "published source differs from validated node"
            )
        parent = self.workspace.nodes.read(identity=pending.parent_node)
        repository = GitRepository(
            io=CommandRunner(descriptors=metadata.descriptors),
            path=self.workspace.config.git_directory,
        )
        projection = GitProjection(repository)
        projection.set_ref(
            reference=f"refs/cowtree/tips/{self.workspace.config.initial}",
            commit=node.git_commit,
            expected=parent.git_commit,
        )
        receipts = ()
        if pending.batch is not None:
            receipts = tuple(
                metadata.call(
                    Operation.RESULT, {"request": member.request.model_dump(mode="json")}
                ).decode("result", TypeAdapter(Receipt))
                for member in pending.batch.members
            )
            if any(
                item.version != receipt.version or item.root != receipt.root for item in receipts
            ):
                raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, "batch receipt mismatch")
        current = Workspace.open(root=self.workspace.root)
        write_record(
            path=self.workspace.root / "receipts" / f"{receipt.version}.json",
            record=PublicationRecord(
                receipt=receipt, validation=pending.validation, receipts=receipts
            ),
        )
        write_record(
            path=self.workspace.root / "workspace.json",
            record=current.config.model_copy(update={"warm_tip": node.id}),
        )
        origins = dict(leaf.origins)
        for path, value in pending.changes.items():
            if value is None:
                origins.pop(path, None)
            else:
                origins[path] = value
        grants = metadata.call(Operation.GRANTS).decode("grants", TypeAdapter(list[Grant]))
        updated = leaf.model_copy(
            update={
                "origins": origins,
                "grants": {grant.path: grant for grant in grants if grant.leaf == leaf.id},
                "sequence": pending.request.sequence + 1,
                "pending": None,
                "last_receipt": receipt,
            }
        )
        Leaves(self.workspace).save(leaf=updated)

    def result(self, request: Request) -> Receipt | None:
        with self.workspace.session() as metadata:
            result = metadata.call(
                Operation.RESULT, {"request": request.model_dump(mode="json")}
            ).decode("result", TypeAdapter(Receipt | None))
        return result

    def recover_locked(self, metadata: Metadata) -> None:
        for leaf in Leaves(self.workspace).records():
            pending = leaf.pending
            if pending is None or not pending.submitted or pending.aborting:
                continue
            receipt = metadata.call(
                Operation.RESULT, {"request": pending.request.model_dump(mode="json")}
            ).decode("result", TypeAdapter(Receipt | None))
            if receipt is not None:
                self.finish(metadata=metadata, leaf=leaf, receipt=receipt)

File: crates/workspace/src/publications.rs
// Copyright (c) 2026 Windsor Nguyen

//! Publish only a checked candidate and recover its exact authority acknowledgement.

use std::fs;

use cowtree_metadata::{Candidate, Receipt, RequestId, Snapshot, objects::ObjectId};

use crate::{
    Error, Leaf, Result, Workspace, error::Issue, records, session::Session,
    types::PublicationRecord, views::replace_origin,
};

impl Workspace {
    pub fn pending_candidate(&self, identity: cowtree_metadata::LeafId) -> Result<Candidate> {
        let leaf = self.lock()?.read_leaf(identity)?;
        leaf.pending
            .and_then(|pending| pending.candidate)
            .ok_or_else(|| Issue::InvalidCandidate.at(&leaf.path))
    }

    pub fn prepare(&self, identity: cowtree_metadata::LeafId) -> Result<Candidate> {
        let mut session = self.lock()?;
        let mut leaf = session.read_leaf(identity)?;
        let pending = leaf
            .pending
            .as_mut()
            .filter(|pending| pending.submitted && !pending.aborting)
            .ok_or_else(|| Issue::PendingPublication.at(&leaf.path))?;
        let candidate = session.authority.prepare(pending.request)?;
        let parent = session.nodes().read(&session.workspace.config.warm_tip)?;
        let (version, entries) = session.published()?;
        if version != candidate.parent || entries != parent.source {
            return Err(Issue::NodeChanged.at(&self.root));
        }
        pending.candidate = Some(candidate.clone());
        pending.batch = None;
        pending.parent_node = Some(parent.id);
        pending.validation = None;
        session.save_leaf(&leaf)?;
        Ok(candidate)
    }

    pub fn commit(&self, candidate: &Candidate) -> Result<Receipt> {
        let mut session = self.lock()?;
        let leaf = session.read_leaf(candidate.request.leaf)?;
        if leaf.pending.is_none() {
            if !leaf
                .last_receipt
                .as_ref()
                .is_some_and(|receipt| receipt.request == candidate.request)
            {
                return Err(Issue::InvalidCandidate.at(&leaf.path));
            }
            return Ok(session.authority.commit(candidate.clone())?);
        }
        let pending =
            leaf.pending.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        let validation = pending
            .validation
            .as_ref()
            .filter(|validation| {
                !pending.aborting
                    && pending.candidate.as_ref() == Some(candidate)
                    && validation.candidate == *candidate
            })
            .ok_or_else(|| Issue::InvalidCandidate.at(&leaf.path))?;
        let node = session.nodes().read(&validation.node)?;
        session.nodes().verify(&node)?;
        if node.source != session.candidate_manifest(candidate)? {
            return Err(Issue::NodeChanged.at(&self.root));
        }
        let receipt = session.authority.commit(candidate.clone())?;
        crate::fault::checkpoint("after-publication-commit");
        session.finish_publication(leaf, &receipt)?;
        session.pin_origins()?;
        Ok(receipt)
    }

    pub fn result(&self, request: RequestId) -> Result<Option<Receipt>> {
        Ok(self.lock()?.authority.result(request)?)
    }
}

impl Session {
    pub(crate) fn candidate_manifest(&self, candidate: &Candidate) -> Result<Snapshot> {
        let path = self.authority.object_directory().join(candidate.root.as_str());
        let bytes = fs::read(&path).map_err(|error| Error::io(&path, error))?;
        if ObjectId::from_bytes(&bytes) != candidate.root {
            return Err(Issue::CorruptObject.at(&path));
        }
        serde_json::from_slice(&bytes).map_err(|error| Error::record(&path, error))
    }

    pub fn finish_publication(&mut self, mut leaf: Leaf, receipt: &Receipt) -> Result<()> {
        let pending =
            leaf.pending.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        let validation =
            pending.validation.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        let candidate =
            pending.candidate.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        let parent_id =
            pending.parent_node.as_ref().ok_or_else(|| Issue::IncompleteRecord.at(&leaf.path))?;
        if receipt.request != pending.request || receipt.root != candidate.root {
            return Err(Issue::InvalidCandidate.at(&leaf.path));
        }
        let (version, entries) = self.published()?;
        if version != receipt.version {
            return Err(Issue::InvalidCandidate.at(&leaf.path));
        }
        let node = self.nodes().read(&validation.node)?;
        if node.source != entries {
            return Err(Issue::NodeChanged.at(&self.workspace.root));
        }
        let parent = self.nodes().read(parent_id)?;
        self.repository().set_ref(
            &format!("refs/cowtree/tips/{}", self.workspace.config.initial.as_str()),
            &node.git_commit,
            Some(&parent.git_commit),
        )?;
        let receipts = self.batch_receipts(pending, receipt, &leaf.path)?;
        let record = PublicationRecord {
            receipt: receipt.clone(),
            validation: validation.clone(),
            receipts,
        };
        records::write(
            &self.workspace.root.join("receipts").join(format!("{}.json", receipt.version.get())),
            &record,
        )?;
        self.workspace.config.warm_tip = node.id;
        records::write(&self.workspace.root.join("workspace.json"), &self.workspace.config)?;
        for (path, value) in &pending.changes {
            replace_origin(&mut leaf.origins, path, value.as_ref());
        }
        leaf.grants = self
            .authority
            .grants()?
            .into_iter()
            .filter(|grant| grant.leaf == leaf.id)
            .map(|grant| (grant.path.clone(), grant))
            .collect();
        leaf.sequence = std::num::NonZeroU64::new(
            pending
                .request
                .sequence
                .checked_add(1)
                .ok_or(cowtree_metadata::Error::CounterExhausted)?,
        )
        .ok_or(cowtree_metadata::Error::CounterExhausted)?;
        leaf.pending = None;
        leaf.last_receipt = Some(receipt.clone());
        self.save_leaf(&leaf)
    }

    fn batch_receipts(
        &self,
        pending: &crate::Pending,
        receipt: &Receipt,
        path: &std::path::Path,
    ) -> Result<Vec<Receipt>> {
        let mut receipts = Vec::new();
        if let Some(batch) = &pending.batch {
            for member in &batch.members {
                let result = self
                    .authority
                    .result(member.request)?
                    .ok_or_else(|| Issue::IncompleteRecord.at(path))?;
                if result.version != receipt.version || result.root != receipt.root {
                    return Err(Issue::InvalidCandidate.at(path));
                }
                receipts.push(result);
            }
        }
        Ok(receipts)
    }

    pub fn recover_publications(&mut self) -> Result<()> {
        for leaf in self.leaves()? {
            if let Some(pending) = &leaf.pending {
                if pending.submitted && !pending.aborting {
                    if let Some(receipt) = self.authority.result(pending.request)? {
                        self.finish_publication(leaf, &receipt)?;
                    }
                }
            }
        }
        Ok(())
    }
}
